### DISCUSSION TITLE
Support acess to class methods(dynamically)

### DISCUSSION THREAD

**Discussion by delciocapolo** [prisma/prisma#21308] [Q&A]
## Problem

A feature that would suit Prisma very well is support for the methods of the PrismaClient class, dynamically. Follow the following example:
`typescript
import {PrismaClient} from "@prisma/client";

const prisma = new PrismaClient();

const getValues = async (tbl_name: string) => {
const rows = await prisma[tbl_name].findMany();

console.log(rows);
}

getValues();
`
This way, there is always an error, stating that **being the string type, cannot be used as an index...**

## Alternative one (actually)
`typescript
import {PrismaClient} from "@prisma/client";

const prisma = new PrismaClient();

const getValues = async () => {

const rows = await prisma.users.findMany();

console.log(rows);
}
`
I tirelessly searched for an answer and still, without success. If there is any solution (using typescript), it would help a lot. Thanks in advance.

**Comment by aqrln** [comment]
Use a narrower type rather than a `string` and restrict the parameter of `getValues` to the valid model property names using a union type. For example, if you have tables `A`, `B` and `C`, then it might look like this:

```diff
- const getValues = async (tbl_name: string) => {
+ const getValues = async (tbl_name: 'a' | 'b' | 'c') => {
    const rows = await prisma[tbl_name].findMany();
```

If you don't want to write them all by yourself explicitly, you can also use this:

```diff
- import {PrismaClient} from "@prisma/client";
+ import { Prisma, PrismaClient } from "@prisma/client";

- const getValues = async (tbl_name: string) => {
+ const getValues = async (tbl_name: Prisma.TypeMap['meta']['modelProps']) => {
    const rows = await prisma[tbl_name].findMany();
```

**Reply by wjw020206** [reply]
Although the type of `prisma[tbl_name]` does not report an error, the `findMany()` method reports an error.

**Comment by delciocapolo** [comment]
When I try to follow the suggestion, I get this error:

**This expression cannot be called.**
  Each member of the union type '(<T extends ProductCategoryFindManyArgs<DefaultArgs>>(args?: SelectSubset<T, ProductCategoryFindManyArgs<DefaultArgs>> | undefined) => PrismaPromise<...>) | ...more 4... | (<T extends VendaFindManyArgs<...>>(args?: SelectSubset<...> | undefined) => PrismaPromise<...>)' **has signatures, but none of these signatures are compatible with each other**. _ts(2349)_

I can invoke the method name! But it comes as if it were a property and you can't use it as a method!

**Reply by wjw020206** [reply]
I also encountered this problem, have you found any solution?
